fix save order of sensors v2

info-cppsp opened the pull request:
@flexiondotorg 
So you understand, what happened.

@raveit65 has been bugging me about this issue :P, so I have wasted at least 10 hours to come up with and implement a second solution.
[The original problem](https://github.com/mate-desktop/mate-sensors-applet/issues/46):
The order of the sensor was not saved by msa.
This was a very old problem, code from gnome-sa was not imported.
The idea there was to load the sensors from gsettings and then from the plugins (loaded normally).
This had the side effect, that even zombie sensors (those removed from the system) have been added from gsettings. (I didn't even think of this ofc.)
This is how msa stands right now as I have added that code with heavy rewrites into msa.

This is not necessarily a bad thing, as now it would be possible to have removable sensors (the zombie ones repurposed, so that) - they are checked every 5 mins or more and get back to active service, if they don't return errors any more.

The other solution to the sorting problem is to load the sensors from the plugins, like normal and then sort them based on a saved sort list in gsettings.
I haven't implemented this before, because as you can see now, it is messy.
It can be optimized a bit, but it's the end of the day, so...

There is yet another solution, which would be easy with OOP.
The idea is to refcount the sensors in a way, where it is apparent, when a sensor is only in gsettings, but no longer available in the system, and then remove it.
(first load from gsettings, with sort, then normal load from plugins, then remove)

NOTE that this commit is working so far, but needs more work.
I shared it, bc it is an algorithm for sorting a tree store and it may or may not be 100% and there may be a much better ways to do this. Please share any ideas.
I haven't measured how long it takes, but there is no increase in memory so far, so hopefully at least it is not leaking...

If you / the team would go through the code and check if it is (at least) logically correct.

raveit65 commented:
I just test this with 2 usb drives.
Unplug one or both removes the drive from sensor list, after the applet or panel was restart.
And non warning popup will be generate.
After i plugged in the drives again the sensors are still there if i restart the applet.
And it seems the order of the sensors are saved too.
So, the behaviour is like the state before we started to save the sensor order.
....good.

Code testing needs to be done by other team members.

info-cppsp commented:
I realized that I am not handling the cases:
1. !found_a
2. found_a && !found_b
I'll make some adjustments later.

info-cppsp commented:
cleaned up comments
the sensors-applet-settings.c file as a whole
added the above 2 cases

This version should be working.
I'll add some optimization code later, that checks the loaded list of sensors against the list from gsettings, so that the sorting doesn't run unnecessarily...

raveit65 commented:
All with 2 usb drives.
Removing one or two drives works well.
After a restart the sensor is gone from the panel, but not from prefs-dialog an gesettings.
Which is Ok for case if you want to plugin the disk again, in result the saved order of the sensor-list is used again.
Well, is possible to delete the sensor complete with dconf-editor.
I am a bit ill, so that is all for the moment.

info-cppsp commented:
added optimization to avoid sorting if the sorting hasn't been changed from normal loading of the sensors
Now this PR should be complete, please check / test.

raveit65 commented:
I noticed some new build warnings.
```
sensors-applet-settings.c: In function 'sensors_applet_settings_sort_sensors_sort':
sensors-applet-settings.c:251:12: warning: variable 'interface_name_b' set but not used [-Wunused-but-set-variable]
     gchar *interface_name_b;
            ^~~~~~~~~~~~~~~~
sensors-applet-settings.c:250:12: warning: variable 'interface_name_a' set but not used [-Wunused-but-set-variable]
     gchar *interface_name_a;
            ^~~~~~~~~~~~~~~~


sensors-applet-settings.c: In function 'sensors_applet_settings_sort_sensors':
sensors-applet-settings.c:528:8: warning: 'ret_val' may be used uninitialized in this function [-Wmaybe-uninitialized]
     if (2 == ret_val) {
        ^
sensors-applet-settings.c:531:5: warning: 'hash_b' may be used uninitialized in this function [-Wmaybe-uninitialized]
     g_free (hash_b);
     ^~~~~~~~~~~~~~~
sensors-applet-settings.c:529:9: warning: 'hash_a' may be used uninitialized in this function [-Wmaybe-uninitialized]
         g_free (hash_a);
         ^~~~~~~~~~~~~~~
```
I think they were there with the first commit.

info-cppsp commented:
removed interface_name_a and interface_name_b
(I planned to use them, but I don't, forgot to remove them.)

added init to NULL to pointers.

I didn't see these warnings, sorry. Please check again.

raveit65 commented:
Great, warnings are gone :)

raveit65 commented:
Removing one or 2 usb drives works well now.
After a restart of the panel th 2 drives aren't any more there as sensor for the panel and they are removed from preferences UI.
But they are still there in dconf settings which might be useful for reading the save order if the usb drive gets plugged in again.
But if i plug in the drives again i see them as sensors in panel again, but they aren't anymore at their previous place.
Well, this isn't a big issue for me (no blocker), but doesn't it make more sense to remove them from dconf too in this case?

@monsta 
Can you do a code review please?

info-cppsp commented:
"they are still there in dconf settings"
ofc as they have been saved. maybe I could remove them, but I don't think that's so important.

"But if i plug in the drives again i see them as sensors in panel again, but they aren't anymore at their previous place."
ofc as if you remove the sensor, the sorting saved won't contain the sensor, so...
that is why I have asked before, whether or not we would want to go this way at all.
this PR (kinda) prevents plug&play...

raveit65 commented:
> that is why I have asked before, whether or not we would want to go this way at all.
this PR (kinda) prevents plug&play

Ok, than is all fine.

raveit65 commented:
>> But if i plug in the drives again i see them as sensors in panel again, but they aren't anymore at their previous place."
> ofc as if you remove the sensor, the sorting saved won't contain the sensor, so...
that is why I have asked before, whether or not we would want to go this way at all.
this PR (kinda) prevents plug&play...

Not sure, if you really understand my intention, so i will answer again.
It is Ok, that the order of an removed drive isn't save, because it is an usb drive.
But i don't see any reason why the removed disk stays in dconf settings.
If someone use his box for testing old disks via usb he will have a lot of unused entries after a while.

Any way, this isn't a blocker for me and the new code improves current situation a lot.
Hope that someone other than me take a look at this and will do a code review, because this belongs in a 1.20 point release.
@vkareh @monsta please.....

info-cppsp commented:
> yoda conditions

they are a good habit in C IMHO ;)

raveit65 commented:
rebased and merged
https://github.com/mate-desktop/mate-sensors-applet/commit/39a1a8c9554f3d027942784159b3a55662afb55e
Thanks for your hard work :)
